plugins/inputs/firehose/firehose.go

```diff
@@ -0,0 +1,413 @@
+	tlsConf, err := f.ServerConfig.TLSConfig()
+	if err != nil {
+		return err
+	}
+
+	var listener net.Listener
+	if tlsConf != nil {
+		listener, err = tls.Listen("tcp", f.ServiceAddress, tlsConf)
+	} else {
+		listener, err = net.Listen("tcp", f.ServiceAddress)
+	}
+	if err != nil {
+		return err
+	}
+	f.tlsConf = tlsConf
+	f.listener = listener
+
+	return nil
+}
+
+func (f *Firehose) ServeHTTP(res http.ResponseWriter, req *http.Request) {
+	select {
+	case <-f.close:
+		res.WriteHeader(http.StatusGone)
+		return
+	default:
+	}
+
+	if !choice.Contains(req.URL.Path, f.Paths) {
+		res.WriteHeader(http.StatusNotFound)
+		return
+	}
+
+	r := &firehoseRequest{req: req}
+
+	if f.AccessKey != "" {
+		if ok := r.authenticate(f.AccessKey); !ok {
+			f.Log.Error(formatLog(req.Header.Get(requestIDHeader), "Unauthorized"))
+			err := r.sendResponse(res)
+			if err != nil {
+				f.Log.Error(formatLog(req.Header.Get(requestIDHeader), err.Error()))
+			}
+			return
+		}
+	}
+
+	if err := r.validate(); err != nil {
+		f.Log.Error(formatLog(req.Header.Get(requestIDHeader), err.Error()))
+		err := r.sendResponse(res)
+		if err != nil {
+			f.Log.Error(formatLog(req.Header.Get(requestIDHeader), err.Error()))
+		}
+		return
+	}
+
+	decodedBytesData, ok := r.decodeData()
+	if !ok {
+		f.Log.Error(formatLog(r.body.RequestID, "Failed to base64 decode record data"))
+		err := r.sendResponse(res)
+		if err != nil {
+			f.Log.Error(formatLog(req.Header.Get(requestIDHeader), err.Error()))
+		}
+		return
+	}
+
+	metrics := make([]telegraf.Metric, 0)
+	for _, bytes := range decodedBytesData {
+		m, err := f.Parse(bytes)
+		if err != nil {
+			f.Log.Error(formatLog(r.body.RequestID, "Unable to parse data"))
+			r.responseStatusCode = http.StatusBadRequest
+			err := r.sendResponse(res)
+			if err != nil {
+				f.Log.Error(formatLog(req.Header.Get(requestIDHeader), err.Error()))
+			}
+			return
+		}
+		metrics = append(metrics, m...)
+	}
+
+	if len(metrics) == 0 {
+		once.Do(func() {
+			f.Log.Info(formatLog(r.body.RequestID, internal.NoMetricsCreatedMsg))
+		})
+	}
+
+	commonAttributesHeaderValue := req.Header.Get(commonAttributesHeader)
+	if len(commonAttributesHeaderValue) != 0 && len(f.ParameterTags) != 0 {
+		parameters := make(map[string]interface{})
+		err := json.Unmarshal([]byte(commonAttributesHeaderValue), &parameters)
+		if err != nil {
+			f.Log.Warn(formatLog(r.body.RequestID, commonAttributesHeader+" header's value is not a valid json"))
+		}
+
+		parameters, ok := parameters["commonAttributes"].(map[string]interface{})
+		if !ok {
+			f.Log.Warn(formatLog(r.body.RequestID, "Invalid value for header "+commonAttributesHeader))
+		}
+
+		for _, parameter := range f.ParameterTags {
+			if value, ok := parameters[parameter]; ok {
+				for _, m := range metrics {
+					m.AddTag(parameter, value.(string))
+				}
+			}
+		}
+	}
+
+	for _, m := range metrics {
+		if f.PathTag {
+			m.AddTag(pathTag, req.URL.Path)
+		}
+		f.acc.AddMetric(m)
+	}
```

Comment:
I really would collect **all** tags _first_ and _then_ iterate over the metrics once to add them.

Reply:
Okay. Does the following makes sense:
```
        if f.PathTag {
		for _, m := range metrics {
			m.AddTag(pathTag, req.URL.Path)
		}
	}
	
	for _, m := range metrics {
		f.acc.AddMetric(m)
	}
```

Reply:
What I want is (pseudocode)
```golang
	tags := make(map[string]string, len(f.ParameterTags)+1)
	if f.pathTag {
		tags[f.pathTag] = req.URL.Path
	}
	for _, parameter := range f.ParameterTags {
		if value, found := parameters[parameter]; found {
			if v, ok := value.(string); ok {
				tags[parameter] = v
			}
		}
	}

	for _, m := range metrics {
		for k, v := range tags {
			m.AddTag(k , v)
		}
		f.acc.AddMetric(m)
	}
```

Reply:
Actually `f.PathTag` variable is a boolean. When set to `True`, we add a tag, `firehose_http_path` to each metric with value set to the request path, i.e, `req.URL.Path`. It's unrelated to `f.ParameterTags`. The purpose of this was to allow the plugin to receive request on more than one request paths. So with the following configuration:
```toml
[[inputs.firehose]]
  ## Address and port to host HTTP listener on
  service_address = ":8080"

  ## Paths to listen to.
  paths = ["/telegraf/dev", "/telegraf/stage"]
 ```
our input plugin would accept requests for both of these paths. This would allow us to have a single telegraf server running, serving both our dev and stage environments. If we want, we can also add the prod environment here or have a separate telegraf server running for it. And if the `f.PathTag` is set to `True` we'll add a tag `firehose_http_path` to each metric with it's value set to the actual request path e.g `telegraf/dev`. This is something I thought could be useful and is optional therefore.

Reply:
updated